trie/triedb/pathdb/database.go

```diff
@@ -276,14 +279,21 @@ func (db *Database) Reset(root common.Hash) error {
 	// Mark the disk layer as stale before applying any mutation.
 	db.tree.bottom().markStale()
 
+	// Reset the clean cache before applying mutation to disk.
+	if db.cleans != nil {
+		db.cleans.Reset()
+	}
```

Comment:
Is this concurrency safe though? I mean, the clean cache is outside of the diskLayer lock I presume? If yes, then an eth_call could in theory read and repopulate the read cache while a disk layer mutation is being applied, so we still can end up with false hits.

Reply:
```
// Cache is a fast thread-safe inmemory cache optimized for big number
// of entries.
```

The clean cache itself is thread-safe. 

Regarding the repopulating the clean cache, the disk layer will be tagged as stale so all following reads will be rejected in the first place. Then we reset the clean cache. So after cache.Reset, nothing new will be loaded again.

Reply:
But I agree it's ugly to expose clean cache outside of `diskLayer`, since it's very disk layer specific. I will try to move it into disklayer.